Crash in CommandEncoder::beginRenderPass using depth stencil attachments from GPUTexture instances

mwyrzykowski opened the pull request:
#### 0243c25f6e577b6c4cc56a951a583e2ac5507da0
<pre>
Crash in CommandEncoder::beginRenderPass using depth stencil attachments from GPUTexture instances
<a href="https://bugs.webkit.org/show_bug.cgi?id=302342">https://bugs.webkit.org/show_bug.cgi?id=302342</a>
radar://164499025

Reviewed by Geoffrey Garen.

We unconditionally attempted to clear the depth stencil view, even
when the depth stencil attachment was passed as a texture.

Test: fast/webgpu/nocrash/fuzz-302342.html
* LayoutTests/fast/webgpu/nocrash/fuzz-302342-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-302342.html: Added.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.beginRenderPass(_:)):

Canonical link: <a href="https://commits.webkit.org/303078@main">https://commits.webkit.org/303078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a75071c52ca36f84783991963bf6e217a654897

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130702 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41657 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138126 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82372 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/933eb2bc-1351-4de8-a3c3-4d8b5d0c9cef) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2989 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2866 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99572 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67466 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/83b2323f-8e0e-41fd-92cf-bcd76d870858) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133649 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2162 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117028 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80280 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/18182b1b-db37-4eee-8a35-1b92e09a2ddb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2086 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81379 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110666 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35161 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140603 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2763 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35662 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108077 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2807 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2492 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108011 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27562 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2126 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113370 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55760 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2833 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31795 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2653 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66222 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2854 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2759 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -1183,8 +1183,11 @@ extension WebGPU.CommandEncoder {
         if attachmentsToClear.count != 0 || depthStencilAttachmentToClear != nil {
             let attachment = wgpuGetRenderPassDescriptorDepthStencilAttachment(descriptorSpan)?[0]
             if attachment != nil && depthStencilAttachmentToClear != nil {
```
this should use `if let`, and then no force unwrapping needs to happen, and no potential crashes can happen

mwyrzykowski replied:
We can write the code like so:
```
            if depthStencilAttachmentToClear != nil {
                if let attachment {
                    var textureView: WebGPU.TextureOrTextureView
                    if let view = attachment.view {
                        textureView = WebGPU.TextureOrTextureView(WebGPU.fromAPI(view))
                    } else if let texture = attachment.texture {
                        textureView = WebGPU.TextureOrTextureView(WebGPU.fromAPI(texture))
                    }
                }
                textureView.setPreviouslyCleared()
            }
```

but I'm not sure its more readable. 

Safe crashing is very much preferred over incorrect logic which doesn't crash.

rr-codes replied:
what about just this (inside the depthStencilAttachmentToClear check thing)

```
let textureView = if let view = attachment?.view { 
   WebGPU.TextureOrTextureView(WebGPU.fromAPI(view))
else if let texture = attachment?.texture {
   WebGPU.TextureOrTextureView(WebGPU.fromAPI(texture))
}
 ```
 
 I _strongly_ think that force unwraps should _always_ be avoided, since they essentially suppress the compiler and turn compile time issues into runtime crashes

rr-codes replied:
in fact, you could even do something like this!

```
extension WebGPU.TextureOrTextureView {

init?(attachment: Attachment?) {
 if let view = attachment?.view { 
   self.init(WebGPU.fromAPI(view))
else if let texture = attachment?.texture {
   self.init(WebGPU.fromAPI(texture))
} else {
  return nil
}

}
```

and then you can just do `let textureView = WebGPU.TextureOrTextureView(attachment)`

rr-codes replied:
endless possibilities of making it easily readable and comprehensible while leveraging Swift’s strengths and not going against the compiler :)

pxlcoder replied:
> I strongly think that force unwraps should always be avoided, since they essentially suppress the compiler and turn compile time issues into runtime crashes

+1, force unwrapping is an anti-pattern.

mwyrzykowski replied:
Can you please explain how it's different from dereferencing a RefPtr (specifically calling `RefPtr::operator*()`)?

I agree we should avoid it, but some of the patterns suggest move runtime crashes into silent failures which are much worse.

rr-codes replied:
> Can you please explain how it's different from dereferencing a RefPtr (specifically calling RefPtr::operator*())?

C++ as a language doesn’t have compiler support for optional values, so runtime checking for null-ness and then dereferencing is the best we can do. But Swift _does_ have compile time support for these, so it can and should be leveraged.

mwyrzykowski replied:
There is a very large risk to introducing silent failures with improper `if let` and `guard let` usage which in GPU process code can be worse than crashing. 

I do like the pattern where it makes sense. The argument that it should be used exclusively is not strong and I think subject to discussion on webkit-dev before any formal policies are in place.

rr-codes replied:
> There is a very large risk to introducing silent failures with improper if let and guard let usage

Hm, do you have any concrete examples of how these might introduce silent failures compared to using a force unwrap? In general, going against the compiler and suppressing a form of type checking should usually be avoided if possible.

geoffreygaren replied:
> this should use if let, and then no force unwrapping needs to happen, and no potential crashes can happen

> +1, force unwrapping is an anti-pattern.

Where is the force-unwrapping? I see an optional passed to `WebGPU.TextureOrTextureView`, which accepts an optional and then uses `if let` to unwrap it.

rr-codes replied:
@geoffreygaren this was prior to the latest version of this PR, so this comment thread is no longer directly applicable

mwyrzykowski replied:
> > There is a very large risk to introducing silent failures with improper if let and guard let usage
> 
> Hm, do you have any concrete examples of how these might introduce silent failures compared to using a force unwrap? In general, going against the compiler and suppressing a form of type checking should usually be avoided if possible.

Yes a great example is this bug :) It would have never been detected had we said:
```
if let view = attachment.view {
   view.setPreviouslyCleared()
}
```
it was only detected because we force unwrapped it like:
```
view!.setPreviouslyCleared()
```

I don't really understand the going against the compiler argument when force-unwrapping is a language feature supported b the compiler :) And in this case it led to the discovery of an issue we would have otherwise likely not discovered until a 3rd party reported it, the CTS expanded coverage in this area, or our own tests expanded coverage here

geoffreygaren replied:
I tend to agree with @mwyrzykowski here that `view!` is clearer than `if let view { } else { fatalError() }`. @rr-codes @pxlcoder can you explain why you find the second form clearer? Or were you referring to a general principle that might not apply in this specific case?

mwyrzykowski replied:
We can continue to discuss, but I think this is outside the scope of this PR.

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -94,6 +94,36 @@ public func CommandEncoder_finish_thunk(commandEncoder: WebGPU.CommandEncoder, d
+            return
+        }
+
+        fatalError()
```
I don’t think this should be a fatalError, since isn’t it valid for the passed in `attachment` to be nil, or for both the view and the texture to be nil?

mwyrzykowski replied:
`TextureOrTextureView` expects a non-nil attachment.view or non-nil attachment.texture. Creating a `TextureOrTextureView` without either is a fatal logic error

rr-codes commented on Source/WebGPU/WebGPU/CommandEncoder.swift:
```diff
@@ -883,7 +913,7 @@ extension WebGPU.CommandEncoder {
             for i in 0..<attachments.count {
                 let attachment = attachments[i]
 
-                if attachment.view == nil {
```
if `WebGPU.TextureOrTextureView` can be made into an optional initializer, this can then be 

```
guard let textureOrTextureView = WebGPU.TextureOrTextureView(attachment) else {
continue
}
```

geoffreygaren reviewed (approved):
I'm going to say r+ to get the bug fix in; but I'm open to further discussion about the best way to represent "nil value is possible, but would be fatal".

mwyrzykowski committed:
Crash in CommandEncoder::beginRenderPass using depth stencil attachments from GPUTexture instances
https://bugs.webkit.org/show_bug.cgi?id=302342
radar://164499025

Reviewed by Geoffrey Garen.

We unconditionally attempted to clear the depth stencil view, even
when the depth stencil attachment was passed as a texture.

Test: fast/webgpu/nocrash/fuzz-302342.html
* LayoutTests/fast/webgpu/nocrash/fuzz-302342-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-302342.html: Added.
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginRenderPass):
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.beginRenderPass(_:)):

Canonical link: https://commits.webkit.org/303078@main
